### DISCUSSION TITLE
More issues with hello_world.py

### DISCUSSION THREAD

**Discussion by GideonLombardo** [USC-ACTLab/crazyswarm#610] [Q&A]
Hello,

This question is related to discussion #603 . When I run hello_world.py, the Crazyflie will either randomly fly away from its starting point or flip over. I have read the discussion in #603. I have updated by Cazyradio firmware according the instructions on the configuration page. I get a success message when I run the scan: 

Found Crazyradio with version 99.55.

I have also made sure to update the Crazyflie firmware. 

For reference, I use Vicon Tracker as my motion capture system with unique marker configurations. Currently, I am testing my setup with only one Crazyflie. 

I start the crazyswarm server, and I am able to see the CF pose in RVIZ. I get no error messages, and everything appears to work cleanly. I run python3 hello_world.py, and the CF flips over or flies far away. I get a message in the terminal:

YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.

cfg = yaml.load(crazyflies_yaml)

Warning: No joystick found!

Traceback (most recent call last):

File "/opt/ros/noetic/lib/python3/dist-packages/rospy/impl/tcpros_base.py", line 567, in connect

self.socket.connect((dest_addr, dest_port))

ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

File "/opt/ros/noetic/lib/python3/dist-packages/rospy/impl/tcpros_service.py", line 509, in call

transport.connect(dest_addr, dest_port, service_uri)

File "/opt/ros/noetic/lib/python3/dist-packages/rospy/impl/tcpros_base.py", line 596, in connect

raise TransportInitError(str(e)) #re-raise i/o error

rospy.exceptions.TransportInitError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

File "hello_world.py", line 22, in <module>

main()

File "hello_world.py", line 17, in main

cf.land(targetHeight=0.04, duration=2.5)

File "/home/workstation3/crazyswarm/ros_ws/src/crazyswarm/scripts/pycrazyswarm/crazyflie.py", line 209, in land

self.landService(groupMask, targetHeight, rospy.Duration.from_sec(duration))

File "/opt/ros/noetic/lib/python3/dist-packages/rospy/impl/tcpros_service.py", line 442, in __call__

return self.call(*args, **kwds)

File "/opt/ros/noetic/lib/python3/dist-packages/rospy/impl/tcpros_service.py", line 512, in call

raise ServiceException("unable to connect to service: %s"%e)

rospy.service.ServiceException: unable to connect to service: [Errno 111] Connection refused

I do not know what is going wrong. If there is another file that needs to be adjusted, or if there is more firmware to be updated, please let me know.

Thank you,
Gideon

**Comment by knmcguire** [comment]
Hi! I suggested to #603 to share the changed hover_swarm.launch file, so perhaps you could do the same as well?

There is also this mocap_helper.launch file that only outputs the mocap values, to make sure that you get the right values for that too.

**Reply by GideonLombardo** [reply]
My university is currently on break. As soon as I get back to the lab, I will post it here. In the meantime, I will keep close watch on #603. Thank you for all of your help.

**Reply by flrnhbr1** [reply]
#603 is solved now, maybe you wanna take a look there again.

**Reply by GideonLombardo** [reply]
Hello everyone, my university is now back from break. 

I have kept a close watch on #603, but I do not believe I have any static that would interfere with takeoff.

Here is hover_swarm.launch file:
```
<?xml version="1.0"?>
<launch>
  <arg name="joy_dev" default="/dev/input/js0" />

  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflieTypes.yaml" />
  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflies.yaml" />

  <node pkg="crazyswarm" type="crazyswarm_server" name="crazyswarm_server" output="screen" >
    <rosparam>
      # Logging configuration (Use enable_logging to actually enable logging)
      genericLogTopics: ["log1"]
      genericLogTopicFrequencies: [10]
      genericLogTopic_log1_Variables: ["stateEstimate.x", "ctrltarget.x"]
      # firmware parameters for all drones (use crazyflieTypes.yaml to set per type, or
      # allCrazyflies.yaml to set per drone)
      firmwareParams:
        commander:
          enHighLevel: 1
        stabilizer:
          estimator: 2 # 1: complementary, 2: kalman
          controller: 2 # 1: PID, 2: mellinger
        ring:
          effect: 16 # 6: double spinner, 7: solid color, 16: packetRate
          solidBlue: 255 # if set to solid color
          solidGreen: 0 # if set to solid color
          solidRed: 0 # if set to solid color
          headlightEnable: 0
        locSrv:
          extPosStdDev: 1e-3
          extQuatStdDev: 0.5e-1
        kalman:
          resetEstimation: 1
      # tracking
      motion_capture_type: "vicon" # one of none,vicon,optitrack,optitrack_closed_source,qualisys,vrpn
      object_tracking_type: "motionCapture" # one of motionCapture,libobjecttracker
      send_position_only: False # set to False to send position+orientation; set to True to send position only
      motion_capture_host_name: "192.168.50.110"
      # motion_capture_interface_ip: "192.168.50.110" # optional for optitrack with multiple interfaces
      save_point_clouds: "/dev/null" # set to a valid path to log mocap point cloud binary file.
      print_latency: False
      write_csvs: False
      force_no_cache: False
      enable_parameters: True
      enable_logging: False
      enable_logging_pose: True
    </rosparam>
  </node>

  <node name="joy" pkg="joy" type="joy_node" output="screen">
    <param name="dev" value="$(arg joy_dev)" />
  </node>

  <node pkg="crazyswarm" type="crazyswarm_teleop" name="crazyswarm_teleop" output="screen">
    <param name="csv_file" value="$(find crazyswarm)/launch/figure8_smooth.csv" />
    <param name="timescale" value="0.8" />
  </node>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find crazyswarm)/launch/test.rviz"/>

  <!-- <node pkg="rqt_plot" type="rqt_plot" name="rqt_plot_x" args="/cf2/log1/values[0]"/> -->
  <!-- <node pkg="rqt_plot" type="rqt_plot" name="rqt_plot_roll" args="/cf1/log1/values[2] /cf1/log1/values[3]"/> -->

</launch>
```

This is mocap_helper.launch file:
```
<?xml version="1.0"?>
<launch>

  <node pkg="crazyswarm" type="mocap_helper" name="mocap_helper" output="screen" >
    <rosparam>
      # tracking
      motion_capture_type: "vicon" # one of vicon,optitrack,optitrack_closed_source,qualisys,vrpn
      motion_capture_host_name: "MARTELLIVICON"
      # motion_capture_interface_ip: "192.168.50.110" # optional for optitrack with multiple interfaces
    </rosparam>
  </node>
</launch>
```

For reference, I will also share my hello_world.py file:
```
"""Takeoff-hover-land for one CF. Useful to validate hardware config."""

from pycrazyswarm import Crazyswarm


TAKEOFF_DURATION = 2.5
HOVER_DURATION = 5.0


def main():
    swarm = Crazyswarm()
    timeHelper = swarm.timeHelper
    cf = swarm.allcfs.crazyflies[0]

    cf.takeoff(targetHeight=1.0, duration=TAKEOFF_DURATION)
    timeHelper.sleep(TAKEOFF_DURATION + HOVER_DURATION)
    cf.land(targetHeight=0.04, duration=2.5)
    timeHelper.sleep(TAKEOFF_DURATION)


if __name__ == "__main__":
    main()
```

Thank you all for your patience and help as I try to work through this.

**Reply by whoenig** [reply]
Make sure that your coordinate system matches the one from ROS, e.g., by setting enable_logging_pose to True and moving the CF around manually. If you are unsure about the orientation (positive x needs to be where the antenna is), you can also try to set send_position_only to True.

**Reply by GideonLombardo** [reply]
I have applied these fixes, and I am still getting the same ConnectionRefused Error. I have noticed some different things from my most recent trials however.

My Chooser interface cannot find the firmware versions to flash to the Crazyflie. It also does not let me see the firmware version I am running. I have not edited anything to change where the firmware was built during my install of Crazyswarm, so I am not sure why this is happening. However, I used the cfclient to flash the most recent firmware to the Crazyflie, so I do not believe this is an issue. 

I tried moving the Crazyflie around manually today, and the Crazyswarm Server lost connection after just a short time of motion. This is the same error that occurs when I attempt to run the hello_world script. After a short period of moving, the Crazyswarm Server loses connection to the Crazyflie, but the Crazyflie motors continue to spin the rotors, causing erratic motion. I have attached an image of RVIZ (apologies for the low quality) to confirm that my coordinate system is calibrated properly, and I have shared an additional error message that may shed more light on the issue. At this point, I do not know if the issue lies with the Crazyflie, the CrazyRadio, or my Vicon System.

Error Message:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Unknown rigid body!
[crazyswarm_server-2] process has died [pid 17926, exit code -6, cmd /home/workstation3/crazyswarm/ros_ws/devel/lib/crazyswarm/crazyswarm_server __name:=crazyswarm_server __log:=/home/workstation3/.ros/log/98a050ba-ab04-11ec-ba32-938e0c05f46f/crazyswarm_server-2.log].
log file: /home/workstation3/.ros/log/98a050ba-ab04-11ec-ba32-938e0c05f46f/crazyswarm_server-2*.log
```
![thumbnail_IMG_4065](https://user-images.githubusercontent.com/99747644/159817097-cc85edee-f09f-4524-8015-950d5841dd7b.jpg)

**Reply by GideonLombardo** [reply]
I have run more tests since Friday, and I believe that my Crazyswarm server is terminating connection when the Vicon system has trouble tracking the CF. Is this something that normally happens? Please let me know.

**Reply by knmcguire** [reply]
Yes that could be a reason why it fails. Have you tried the mocap helper launch script?

**Reply by GideonLombardo** [reply]
Here is my mocap_helper.launch file: 
 
```
<?xml version="1.0"?>
<launch>

  <node pkg="crazyswarm" type="mocap_helper" name="mocap_helper" output="screen" >
    <rosparam>
      # tracking
      motion_capture_type: "vicon" # one of vicon,optitrack,optitrack_closed_source,qualisys,vrpn
      motion_capture_host_name: "#######"
      # motion_capture_interface_ip: "######" # optional for optitrack with multiple interfaces
    </rosparam>
  </node>
</launch>
```

Is there a way that I can see motion capture values after trying the hello_world.py script?

My plan is to redefine my Crazyflies in Vicon and hope that helps with the tracking. As far as the axes go, I am still unsure if my coordinate system matches like it is supposed to. 

Another issue I have noticed is that my chooser.py script always says that I have 0.00V, but it is able to do send commands to the crazyflie. Is it possible that the crazyflie is unable to send data back to the Crazyradio?

**Reply by GideonLombardo** [reply]
I have been able to run the hello_world.py script successfully. However, I do have some questions. The script does not work every time. For example, when the crazyflie coordinate axes do not align with the world axes, the script will sometimes fail. The crazyflie axes defined in RVIZ will sometimes shake and flicker. I have a feeling the inconsistency in stable flight is arising from the "shaky" axes. Is there a known reason why these things are happening? Please let me know.

**Reply by aarsht7** [reply]
I am facing the similar issues as you are. I am using `vrpn` with unique markers and it works some times it doesn't most of the time. When it doesn't work, it will just flip over. I tried with `send_position_only = True`. One strange thing happens when I run the `hover_swarm.launch` is, it is logging warning `no updated pose found for the cf#`. But the pose do shown up in rviz. It doesn't flicker much or at least it is not noticeable. I do not know what the problem can be. I have one more thing to ask, if we are using `send_position_only = True` then the flickering of the axis (angular deviation) should not affect the actual behavior of the flight, right?

**Reply by knmcguire** [reply]
@GideonLombardo sorry for the delayed answer, but I'm not sure what is going wrong on your end...

This seems like you probably facing a different issue with this. I would advise you start a new discussion per topic, so that it is easier for others to help you with this.

**Reply by aarsht7** [reply]
I am not sure but it can be the issue of balancing. I found [this](https://www.bitcraze.io/documentation/tutorials/getting-started-with-crazyflie-2-x/#unbalanced) and [this](https://www.bitcraze.io/documentation/tutorials/balancing-propellers/) on their official website. I will try it out before my next testing session and update the flipping issue of the CF.
